lib/internal/modules/esm/loader.js

```diff
@@ -194,34 +194,43 @@ function nextHookFactory(chain, meta, { validateArgs, validateOutput }) {
  * the main module and everything in its dependency graph.
  */
 class ESMLoader {
+  /**
+   * Whether this loader has userland hooks.
+   * @private
+   * @property {boolean} hasLoaderHooks
+   */
+  #hasLoaderHooks = false;
+
   /**
    * Prior to ESM loading. These are called once before any modules are started.
    * @private
-   * @property {KeyedHook[]} globalPreloaders Last-in-first-out
+   * @property {KeyedHook[]} globalPreloadHooks Last-in-first-out
    *  list of preload hooks.
    */
-  #globalPreloaders = [];
+  #globalPreloadHooks = [];
```

Comment:
@guybedford I like the refactoring you did in this PR. Can it be split off into its own PR, without the added `preImport` hook? cc @JakobJingleheimer

Reply:
Oow. `#hasCustomHooks` 🙂 (but actually, I think that's not as useful as `this.#HOOK_NAME.length > 1`?)

FYI, I was about to tidy the hooks up to something like:

```js
#chains = {
  globalPreload: [],
  resolve: [],
  load: [],
}
```